c++/src/kj/string.h

```diff
@@ -451,12 +536,40 @@ auto toCharSequence(T&& value) -> decltype(_::STR * kj::fwd<T>(value)) {
 }
 
 namespace _ {  // private
+template <typename T, typename = void>
+struct IsSignalSafeToCharSequence_ {
```

Comment:
`tryToCharSequence` was just convenient shorthand to avoid needing SFINAE to distinguish if the type is stringifiable in the first place, not because DebugComparison has some special dependency on `tryToCharSequence`. PTAL at the latest code.

Reply:
I'm well aware of what `tryToCharSequence` does, I wrote it after all.

My comment above suggests a strict simplification of the code, allowing `IsSignalSafeToCharSequence_` to be deleted entirely and making the implementation of `isSignalSafeToCharSequence()` much clearer. I tested this suggestion myself and confirmed it worked. Is there a reason you didn't apply it?

Reply:
I'm probably doing something wrong?
```
template <typename T>
static constexpr inline bool isSignalSafeToCharSequence() {
  // Returns true if T is safe to stringify in a signal handler context. This means no memory
  // allocations and no invocations of methods that are not safe not use in a signal handler
  // context. These are manually annotated for each type and this must be done carefully to avoid
  // accidentally allowing types to be passed to strPreallocated. This can be invoked on a type T
  // that isn't itself valid to pass to toCharSequence, in which case this returns false.
  return isInstanceOfTemplate<
      decltype(_::tryToCharSequence(instance<Decay<T>*>())), SignalSafeCharSequence>();
}
```

<details>
<summary>This fails to compile because it seems like it ends up instantiating the wrong thing?</summary><p>
```
In file included from /home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string-test.c++:22:
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:865:7: error: use of deleted function ‘kj::Array<T>::Array(const kj::Array<T>&) [with T = kj::SignalSafeCharSequence<kj::CappedArray<char, 17> >]’
  865 | class Delimited {
      |       ^~~~~~~~~
In file included from /home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/memory.h:24,
                 from /home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/array.h:24,
                 from /home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:25,
                 from /home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string-test.c++:22:
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/array.h:146:3: note: declared here
  146 |   KJ_DISALLOW_COPY(Array);
      |   ^~~~~~~~~~~~~~~~
In file included from /home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string-test.c++:22:
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h: In instantiation of ‘constexpr kj::SignalSafeCharSequence<T>::SignalSafeCharSequence(Args&& ...) [with Args = {const kj::_::Delimited<kj::ArrayPtr<int> >&}; T = kj::_::Delimited<kj::ArrayPtr<int> >]’:
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:961:10:   required from ‘const kj::SignalSafeCharSequence<kj::_::Delimited<T> >& kj::_::operator*(kj::_::Stringifier, const kj::_::Delimited<T>&) [with T = kj::ArrayPtr<int>]’
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:497:17:   required from ‘decltype ((kj::_::STR * fwd<T>(value))) kj::toCharSequence(T&&) [with T = kj::_::Delimited<kj::ArrayPtr<int> >&; decltype ((kj::_::STR * fwd<T>(value))) = const kj::SignalSafeCharSequence<kj::_::Delimited<kj::ArrayPtr<int> > >&]’
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:502:68:   required from ‘auto kj::_::tryToCharSequence(T*) [with T = kj::_::Delimited<kj::ArrayPtr<int> >; <template-parameter-1-2> = const kj::SignalSafeCharSequence<kj::_::Delimited<kj::ArrayPtr<int> > >&]’
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:625:36:   required from ‘constexpr bool kj::isSignalSafeToCharSequence() [with T = kj::_::Delimited<kj::ArrayPtr<int> >]’
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string-test.c++:52:67:   required from here
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:301:77: error: use of deleted function ‘kj::_::Delimited<kj::ArrayPtr<int> >::Delimited(const kj::_::Delimited<kj::ArrayPtr<int> >&)’
  301 |   constexpr SignalSafeCharSequence(Args&&... args): T(kj::fwd<Args>(args)...) {}
      |                                                                             ^
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h: In instantiation of ‘constexpr kj::SignalSafeCharSequence<T>::SignalSafeCharSequence(Args&& ...) [with Args = {const kj::_::Delimited<kj::ArrayPtr<kj::StringPtr> >&}; T = kj::_::Delimited<kj::ArrayPtr<kj::StringPtr> >]’:
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:961:10:   required from ‘const kj::SignalSafeCharSequence<kj::_::Delimited<T> >& kj::_::operator*(kj::_::Stringifier, const kj::_::Delimited<T>&) [with T = kj::ArrayPtr<kj::StringPtr>]’
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:497:17:   required from ‘decltype ((kj::_::STR * fwd<T>(value))) kj::toCharSequence(T&&) [with T = kj::_::Delimited<kj::ArrayPtr<kj::StringPtr> >&; decltype ((kj::_::STR * fwd<T>(value))) = const kj::SignalSafeCharSequence<kj::_::Delimited<kj::ArrayPtr<kj::StringPtr> > >&]’
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:502:68:   required from ‘auto kj::_::tryToCharSequence(T*) [with T = kj::_::Delimited<kj::ArrayPtr<kj::StringPtr> >; <template-parameter-1-2> = const kj::SignalSafeCharSequence<kj::_::Delimited<kj::ArrayPtr<kj::StringPtr> > >&]’
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:625:36:   required from ‘constexpr bool kj::isSignalSafeToCharSequence() [with T = kj::_::Delimited<kj::ArrayPtr<kj::StringPtr> >]’
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string-test.c++:53:73:   required from here
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:301:77: error: use of deleted function ‘kj::_::Delimited<kj::ArrayPtr<kj::StringPtr> >::Delimited(const kj::_::Delimited<kj::ArrayPtr<kj::StringPtr> >&)’
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h: In instantiation of ‘constexpr kj::SignalSafeCharSequence<T>::SignalSafeCharSequence(Args&& ...) [with Args = {const kj::_::Delimited<kj::Array<const kj::Quantity<long int, kj::_::NanosecondLabel>*> >&}; T = kj::_::Delimited<kj::Array<const kj::Quantity<long int, kj::_::NanosecondLabel>*> >]’:
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:961:10:   required from ‘const kj::SignalSafeCharSequence<kj::_::Delimited<T> >& kj::_::operator*(kj::_::Stringifier, const kj::_::Delimited<T>&) [with T = kj::Array<const kj::Quantity<long int, kj::_::NanosecondLabel>*>]’
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:497:17:   required from ‘decltype ((kj::_::STR * fwd<T>(value))) kj::toCharSequence(T&&) [with T = kj::_::Delimited<kj::Array<const kj::Quantity<long int, kj::_::NanosecondLabel>*> >&; decltype ((kj::_::STR * fwd<T>(value))) = const kj::SignalSafeCharSequence<kj::_::Delimited<kj::Array<const kj::Quantity<long int, kj::_::NanosecondLabel>*> > >&]’
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:502:68:   required from ‘auto kj::_::tryToCharSequence(T*) [with T = kj::_::Delimited<kj::Array<const kj::Quantity<long int, kj::_::NanosecondLabel>*> >; <template-parameter-1-2> = const kj::SignalSafeCharSequence<kj::_::Delimited<kj::Array<const kj::Quantity<long int, kj::_::NanosecondLabel>*> > >&]’
/home/vlovich/cloudflare/edgeworker/deps/capnproto/c++/src/kj/string.h:625:36:   required from ‘constexpr bool kj::isSignalSafeToCharSequence() [with T = kj::_::Delimited<kj::Array<const kj::Quantity<long int, kj::_::NanosecondLabel>*> >]’
```
</p>
</details>